### DISCUSSION TITLE
How will RAP be structured?

### DISCUSSION THREAD

**Discussion by greatSumini** [EveryAnalytics/react-analytics-provider#209] [General]
**맥락**
https://github.com/EveryAnalytics/react-analytics-provider/discussions/141

**위클리에서 나온 의견**
1. core -> 리액트 의존 없이 anlaytics 인터페이스 정의
2. react-provider -> anlaytics 인터페이스. 사용한 컴포넌트 제공
3. PA별로 인터페이스 구현체 제공하는 모듈
1번하고 3번만 가져다쓰면 vue나 svelte에서도 쓸 수 있을 것 같아요

아래 두가지 옵션이 있는 것 같습니다!
1. lodash처럼 monolithic하게 만들고, 트리셰이킹으로 해결한다.
2. 아예 패키지를 분리한다.

**Comment by milooy** [comment]
RAP에서 현재 제공하는 기능들 서로 의존 없는 모듈로 쪼개보기:
(위클리에서 말했던것보다 하나 더 쪼갰어요)

1. AnalyticsProvider Context
2. Analytics Components
3. Google Analytics Helper
4. Amplitude Helper

**Reply by milooy** [reply]
1 2 는 react 의존
3 4 는 react 의존 X

**Comment by milooy** [comment]
### A. monolithic + 트리셰이킹
```
yarn add lodash

import array from "lodash/array"
import object from "lodash/fp/object"
```
- 한번만 package install해도 된다. 패키지 버전관리도 하나로.
- 슈퍼앱 너낌... 트리셰이킹 된다 강조해도 부담스러울수 있다

근데 로대쉬 [모듈 쪼갠것](https://www.npmjs.com/search?q=keywords:lodash-modularized)도 제공하네요 ㅋㅋ



### B. 아예 패키지 분리
```
yarn add @emotion/react
yarn add @emotion/styled

import styled from '@emotion/styled'
```
- 패키지별로 버전관리 가능
- 패키지별로 서로 의존성 없음이 명확히 보임

**Reply by sensecodevalue** [reply]
현재까지의 생각입니다. 트리쉐이킹이나 모듈 분리의 개념을 더 읽고 업데이트를 진행하겠습니다!
A. 
- A 같은 경우는 저처럼 약간 아기개발자 같은 사람들에게는 확실이 더 편할 것이라는 생각이 듭니다!
- 하지만 각각을 Import해서 사용할 것이라면 그냥 B 처럼 패키지를 분리하는 것이 더 명확해서 좋을 수 있겠다 생각합니다!

B.  
- 패키지를 쪼갠다면 우리의 패키지를 사용할때 install을 여러개를 진행하게 되는건가요?
- 만약 그렇다면 저같은 아기개발자를 위해서 설명이 충분히 진행 된다면 B방법이 더 뚜렸해서 쓰기 좋을 것 같아요!

**Comment by milooy** [comment]
디스커션 듀데잇 정해볼까요~? 
월/화?

**Reply by sensecodevalue** [reply]
화요일에 하고 싶습니다!

**Reply by josworks27** [reply]
저도 화요일 좋습니다!

**Reply by uoon-dev** [reply]
화요일 좋습니다!

**Comment by josworks27** [comment]
저는 monolithic 하게 하는 것 보다 B번, 패키지를 분리하는 편이 좋을 것 같습니다.
먼저 monolithic하게 했을 때, 개인적으로 트리셰이킹이 될 수 있도록 고려해서 세팅/사용하는 것 보단 모듈 분리를 통해 트리셰이킹이 필요 없도록 하는 편이 좋지 않을까 싶습니다.(물론 각 모듈 내에서의 기능들은 트리셰이킹이 되도록 세팅)

그리고 B번, 패키지를 분리하는 방법으로 했을 땐, RAP가 react에서의 사용을 기반으로 만들어진 라이브러리라는 것을 고려하여 아래와 같은 구조를 제안드립니다. (패키지의 네이밍은 논의가 필요합니다)

먼저 모듈의 분류는 @milooy 님의 분류 기준을 따릅니다.
1. AnalyticsProvider Context
2. Analytics Components
3. Google Analytics Helper
4. Amplitude Helper
5. 기타 utils (*Optional)

`@rap/react`: 1 + 2 + 3 + 4
- 모든 기능을 리액트에서 즉시 사용할 수 있도록 구성

`@rap/core`: 1 + 3 + 4
- 컴포넌트를 제외한 라이브러리의 핵심적인 기능을 즉시 사용할 수 있도록 구성

`@rap/helper`: 3 + 4
- ga/amplitude helper만 제공 (다른 프레임워크에서 사용 가능)

`@rap/utils`: 5
- isSSR 등의 유틸이 더 추가된다면 별도로 분리
- 유저에게 편의성을 높이는 방향으로 갔을 때 분리 필요성이 발생할 것으로 생각

그리고 각 패키지는 모노레포를 통해 하나의 레파지토리 내 packages에서 관리, 각 패키지는 peerDependencies를 통해 각 패키지에서 필요한 의존성을 관리하는 방법으로 가는 편이 좋지 않을까하는 개인적인 의견입니다~!

**Reply by greatSumini** [reply]
음 저는 나눠진 패키지들끼리 직교성이 보장되면 좋을 것 같습니다! 서로 겹치는게 없는 고유한? 관계로요 ㅎㅎ
만약 서로 다른 이름의 패키지가 공통된 코드 베이스를 공유하고 있다면 온보딩하는 입장에서 헷갈릴 수 있을 것 같슴다

**Reply by josworks27** [reply]
@greatSumini 목적에 따라 명확히 분리가 가능하면 분리하는게 좋을 것 같네요 😄
그러면, 현재 구조에서 각 기능별로 분류하고 다른 분류와는 디커플링시키는 작업이 필요하고, 디커플링된 분류가 하나의 패키지를 구성하게 되겠네요

**Reply by greatSumini** [reply]
넵 현재도 react 관련 코드를 제외하고는 커플링이 아예 없어서 ㅎㅎ 분리하기 쉬울 것 같습니다!

**Reply by uoon-dev** [reply]
저도 목적에 따라 명확히 분리하는 방향에 적극 찬성입니다 :)

**Comment by greatSumini** [comment]
저도 나누는 편이 좋을 것 같슴다
릴리즈 템포를 각자 사정에 맞춰서 가져갈 수 있어서 좋을 것 같아요 ㅎㅎ

**Comment by milooy** [comment]
B. 패키지분리 쪽으로 의견이 좁혀졌네요 ㅎㅎ
그렇게 가봅시다!

이제 모듈별로 이름을 지어줘야하는데..의견 부탁드립니다 ㅋㅋ

**Reply by greatSumini** [reply]
`@every-analytics/react-provider` or `@every-analytics/provider` or `@every-analytics/react`
`@every-analytics/google-analytics`
`@every-analytics/amplitude`

vue, svelte에서도 사용할 수 있겠다는 의견을 듣고 생각해봤습니다 ㅎㅎ

**Reply by sensecodevalue** [reply]
저는 좋습니다!!
@every-analytics/react-provider or @every-analytics/provider or @every-analytics/react
셋 중 고른다면 @every-analytics/react-provider 이게 가장 좋습니다.

**Reply by wplong11** [reply]
@greatSumini 리액트 의존 없는 코어 모듈이 @every-analytics/provider에 들어간다고 보면 될까요?

**Reply by wplong11** [reply]
저는 이렇게 생각했었어요.

@every-analytics/every-analytics or @every-analytics/every-analytics.ts or @every-analytics/core
@every-analytics/google-analytics
@every-analytics/amplitude
@every-analytics/react

+) @every-analytics/every-analytics.ts를 생각한 이유는 kotlin / swift 구현도 이 organization에 넣어보고 싶었기 때문입니다

**Reply by uoon-dev** [reply]
@greatSumini @wplong11 패키지별 역할에 헷갈리는 부분이 있어서 질문 드립니다! react에 의존성이 없는 pa util 함수가 @every-analytics/google-analytics, @every-analytics/amplitude 패키지로 분리된다면 @every-analytics/react는 provider와 hook을 포함하는 방향이지 않을까 싶은데 맞을까요~? 기본 기능(initialize, page view, event 등)과 확장 기능(impression 등)도 합쳐지게 되구요. 스레드에서 논의되는 @every-analytics/core 패키지는 공통으로 쓰이는 유틸을 작성하는 패키지가 아니라 기본(핵심) 기능을 지원하는 패키지의 컨텍스트로 읽혀지는데, 이 경우 확장 기능은 위에 다른 패키지에서 정의하는 방향인 건지 조금 헷갈리네요 ㅎㅎ ㅠ 저는 패키지 간의 직교성과 활용성을 고려하면 @every-analytics/google-analytics, @every-analytics/amplitude, @every-analytics/react 세 개로 나누면 되지 않을까 싶었거든요.

+취향이지만 @emotion/react 처럼 @every-analytics/react 패키지명 좋은 것 같아요 ㅎㅎ

**Reply by wplong11** [reply]
> @every-analytics/react는 provider와 hook을 포함하는 방향이지 않을까 싶은데 맞을까요~? 기본 기능(initialize, page view, event 등)과 확장 기능(impression 등)도 합쳐지게 되구요.

넵!

> @every-analytics/core 패키지는 공통으로 쓰이는 유틸을 작성하는 패키지가 아니라 기본(핵심) 기능을 지원하는 패키지의 컨텍스트로 읽혀지는데

저는 #211 PR의 Analytics가 Core 모듈에 들어가는 걸로 생각하고 있어요! PA별 모듈은 Core 모듈을 참조해서 Core에 정의된 Analytics 인터페이스의 구현체를 제공하고, react를 안 쓰는 프로젝트에선(바닐라js라고 부르면 될까요?) PA별 모듈만 참조해서 쓰면 될 것 같아요. 어떤 개발자는 Core 모듈만 참조하고 PA별 구현체는 직접 만들수도 있겠죠! 아니면 자체 PA 도구를 쓴다거나, 하이브리드 앱 지원이 필요하다면 구현체를 직접 정의할 수도 있고요

**Reply by uoon-dev** [reply]
아하 core 모듈은 인터페이스만 제공해주는 역할인가 보군요 :)

**Reply by wplong11** [reply]
@uoon-dev #211 PR 본문에서 소개된 CompositeAnalytics도 들어갈 수 있습니다! 정확히 어떤 것들이 들어간다라고 말로 잘 정리가 안 되네요. 좀 더 잘 말 해보도록 고민해볼게오 😢

아, 그리고 이 내용은 제가 생각하는 방향일 뿐이고 다른 분들은 어떻게 생각하실지 궁금하네요!

**Reply by greatSumini** [reply]
@wplong11 오오 PR 확인해보겠습니다!!

**Comment by milooy** [comment]
Todo도 같이 얘기해봐요

- [ ] 패키지별 이름짓기
- [ ] 라이브러리 구조 설계 - 모노레포?
- [ ] 좍좍 찢기

**Reply by uoon-dev** [reply]
- [ ]  패키지별 README 작성

**Reply by sensecodevalue** [reply]
- [ ] 가이드 작성?! (install 및 각 패키지 간단 사용방법)

**Reply by wplong11** [reply]
- [ ] yarn workspace?

**Comment by wplong11** [comment]
Product Analytics 별로 구현될 모듈과 core 모듈 방향에 영향 미칠 #211 PR을 하나 작성했습니다.

**Comment by uoon-dev** [comment]
저희가 여기 디스커션에서 패키지를 분리하고 별도의 릴리즈 템포를 가져가는 걸로 얘기가 진행이 되었었는데요..! 
어제 회의를 하다 보니 별도의 히스토리를 관리하는 작업이 개발 공수가 더 들기도 하고 작업에 딜레이 되는 감이 없잖아 있는 것 같아서 
우선 빠른 배포와 개발을 위해 모놀리틱 구조로 진행을 하고 패키지 분리 작업은 이후에 개선사항으로 작업하는 게 어떨까 하는 의견이 있었습니다. (@hiphapis @wplong11 @uoon-dev) 레포 구조만 잘 구성한다면 패키지 분리는 이후에 개발되어도 괜찮을 것이라는 판단(+ 애자일,.ㅎ)입니다.
인터페이스와 패키지 코드에 대해선 동현 님이 상세히 설명해주시겠지만 아래 코드처럼 가면 좋을 것 같다고 의논을 하게 되었어요.

```js
import Analytics from "every-anlytics" // for JS
import {useAnalyticsContex} from "every-anlytics/react" // for React

// 추상화된 공통된 인터페이스 예제
Analytics.init(gaKey, amptKey) => {
	Analytics.ga.init(gaKey)
	Analytics.ampt.init(amptKey)
}

Analytics.click('') {
  Analytics.ga.click
  Analytics.ampt.click
}

// PA별로 adaptor가 있다
Analytics.ampt.init("")
Analytics.ampt.revenue({})
Analytics.ampt.click()
```

이미 디스커션에서 패키지 분리 방향으로 결정되어서 조심스럽긴 한데 지금은 빠른 개발에 우선순위가 좀 더 있는 것 같다는 의견에 
길드원 분들의 공감이 여럿 있어서 부득이 (찐)마지막으로 의견을 다시 구하고 싶습니다. 🙏
@milooy @greatSumini @josworks27 @SenseCodeValue

**Reply by greatSumini** [reply]
일단 빠르게 런칭하는 것 좋은 것 같습니다!! 혹시 `every-analytics/react` 패키지는 따로 관리되는 것인가요??

**Reply by wplong11** [reply]
@greatSumini 패키지를 따로 관리하는 것은 아니에요. 패키지 분리를 염두해서 폴더 구조와 폴더 간 의존 관계 합의를 하고, 여유 되면 패키지 분리까지 진행하는 것으로 이야기했어요

**Comment by milooy** [comment]
# 10/12 화요일까지 Todo 
- 수민, 요한 유림 각자 생각하는 짬뽕형 interface 적기
https://www.notion.so/508c29e3fc5b464aa8ae95b4a48e361e

**Reply by greatSumini** [reply]
PR로 업로드해봤습니다 ㅎㅎ

https://github.com/EveryAnalytics/react-analytics-provider/pull/225

**Reply by milooy** [reply]
저도 PR로 업로드 했어요!
https://github.com/EveryAnalytics/react-analytics-provider/pull/226

**Reply by hiphapis** [reply]
(늦어서 말은 못하고) 휙 https://github.com/EveryAnalytics/react-analytics-provider/pull/228

**Comment by milooy** [comment]
## 10/10 회의 코드제안 by @greatSumini , @hiphapis 

### 요구사항들

1. 리액트와의 의존성이 최소화되면 좋다.
2. 여러 PA를 무리 없이 지원해야만 한다.
    
    (적어도 ga, amplitude는 포함해야한다.)
    
3. 공통의 인터페이스가 모든 PA의 모든 기능들을 cover할 필요는 없다.

### 기존 솔루션 분석

공통 interface를 구현한 각 Analytics instance들을 Composite패턴으로 조합해서 사용. (참고: 동현님 PR)

***뭐가 문제일까?***

1. 자바스크립트/프론트엔드 개발 생태계에서 아직 Composite패턴이 익숙하지 않다.
2. 각 PA가 다른 parameter를 요구할텐데, 이를 반영하지 않고 있다. (= Practical하지 않다)

## 새로운 패턴 제안

### 1. 빌더 패턴 (사용자 입장)

```tsx
import { EveryAnalyticsBuilder } from '@every-analytics/core';

interface IAnalytics {
  click: (~~) => ~~;
  setUser: (~~) => ~~;
  init: (~~) => ~~;
  // 그 외 공통 함수들
}

class GoogleAnalytics implements IAnalytics {
}

class Amaplit

class AnalyticsManager implements IAnalytics {
  products = [];

  click() {
    this.products.forEach(product => product.click());
  }
}

// 초기화
const myAppAnalytics = 
  EveryAnalyticsBuilder
    .addGoogleAnalytics({ trackingId: '' })
    .addAmplitude({ apiKey: '' })
    .build(); // 유림: 필요한가융? 수민: 빌더패턴에선 있는게 익숙하다 느껴서 넣었는데 필요없다면 build 네이밍 자체를 빼도 되겠습니당

// 사용 - 클릭
myAppAnalytics.click();
// PA별 기능 활용
myAppAnalutics.googleAnaltics.logCheckout(~~);
myAppAnalytics.amplitude.logRevenueV2(~~);
```

```tsx
// 요한
import { ea } from '@every-analytics';

ea.init({
ua_key: process.env.ua_key,
g4: '~~~',
amplitide: '~~'
})

ea.click()
```

- 수민: init 이 유저 입장에서는 더 익숙하지만 builder패턴이 더 확장성있진 않을까?
    - 그래서 유저입장에서는 요한님방식으로, RAP레벨에서는 builder패턴으로 하면 어떨까?
- 유림: PA툴들이 IAnalytics 에 의존성이 생기는게 아쉽지 않나?
    - 회사로깅을 넣고싶다면 어떻게해야할까융?
    - 수민: 현재 위 코드에선 안되는데 동현님이 제안해주신 composit패턴에선 가능해요
    - 요한: 그럴꺼면 굳이 RAP를 쓸필요가 있나?

**Comment by milooy** [comment]
## 10/10 회의 코드제안 by @milooy 

# 진유림

EveryAnalytics/core

- 함수 한번 호출로 미리 정해둔 여러 PA코드를 호출할 수 있다.

```jsx
import Analytics from "every-analytics/core"
import googleAnalyticsHelper from "EveryAnalytics/helper/google-analytics"
import amplitudeHelper from "EveryAnalytics/helper/amplitude"

Analytics.init(() => {
  googleAnalyticsHelper.initialize(gaid)
  amplitudeHelper.initialize(amtid);
  fruitLogger.init();
});

Analytics.preset({
	click: (name, params) => {
		googleAnalyticsHelper.click(name, params)
		amplitudeHelper.logEvent(amtid, params);
	},
});

// Analytics.init();
button.addEventListener("click", () => {
  Analytics.click('회원가입', {name: 'RAP'});
});
```

- 요한: init같은경우는 preset안에 있는게 아니고 따로 제공해줘야 하지 않을까? (타입선언 강제 등) 혹은 preset이랑 같은 레벨로 init이 있는게 좋지 않을까
- 유림: 아예 beforeAll로 해서 init()과정을 빼버리는건 어떨까?
- 요한: 요한수민유림 짬뽕해서 defalut preset + custom preset 해보는건 어떨까?
- 수민: 유림 레벨정도의 추상화가 현재 상황에서 적절하지 않을까?

EveryAnalytics/react

- React Context, Logging Components 제공

EveryAnalytics/helper/google-analytics

EveryAnalytics/helper/amplitude

**Comment by milooy** [comment]
# 1014 회의
참석자: 요한, 수민, 다영

## 액션아이템
- everyAnalytics/core 의 이름 변경: core라고 하기엔 하는 일이 크지 않다 - 차라리 mixin등 다른 좀더 라이트하고 내용을 설명하는 이름은 어떨까?: 가제로 mixin 땅땅땅
- 모듈 분리. 안 되면 폴더라도 분리

## 엔드유저 입장
- helper들만 쓰는 사람들도 많을 것 같아!
- 다영: 나라면 helper부터 쓸거같다. 그러다가 PA툴 여러개 쓸때 되어서나 core쓸듯. 뭘 도와주는지 나에게 크게 와닿지 않는당.
- 요한: 외부개발자라면 core만 무적권 쓰지 않을까? helper만 따로 쓸수있는게 헷갈리지 않을까? 이렇게도 쓸수있고 저렇게도 쓸 수 있어^^ 라고 하면 ㅠㅠ. 강력한 컨벤션으로 하나만 알려주는게 좋을듯.
- 다영: 자유도와 미리정해져있는 편함과의 trade off가 있는것같아용
- 요한: 어떻게 이걸 설득력있게 전달할지? 여러개가 있으면 헷갈린다.
- 다영: README에 '너는 GA를 원하니? ㅁ를 써' '너는 ampl를 원하니? ㅁ를 써'. '너는 둘다 원하니? mixin을 써!'그리고 라이브러리 이름이 설득력있다면 개선될것같다.

## 새로 개발해야하는것
**재설계**
1. 유림, 요한 좋은 인터페이스 짬뽕해다가 수민님꺼에 합체 @수민

**기능더하기**
2. preset 호출시 바로 init까지 호출 @수민
3. initialize async콜 대응 @수민

**데모**
4. 작동하는 demo 만들기 @수민님끝나고다시나누기

**추가작업**
5. 킬러피쳐: simple version initialize 고찰필요 @유림 @다영 @요한
6. addFunctions @유림 @다영 @요한
7. gaHelper.init랑 amptHelper.init에도 비동기 지원 @유림

**Reply by milooy** [reply]
simple version initialize 예시 추가해봤습니다~
https://github.com/EveryAnalytics/react-analytics-provider/pull/226/commits/36d6417cdd31ed4fe189e6d2a83d76bb0d440c89
